codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AssembleMiddlewareStack.java

```diff
@@ -78,28 +86,67 @@ private void writeAddMiddlewareBefore(
     ) {
         String st = String.format("%s.%s.Add(%s, middleware.Before)", stackOperand, stackstep, content);
         writer.write(st, middlewareSymbol);
-        writer.addUseImports(SmithyGoDependency.SMITHY_MIDDLEWARE);
     }
 
     @Override
     public List<RuntimeClientPlugin> getClientPlugins() {
         return ListUtils.of(
+                // Add ServiceMetadataProvider to operation stack
+                RuntimeClientPlugin.builder()
+                        .buildMiddlewareStack((writer, service, operation, protocolGenerator, stackOperand) -> {
+                            Symbol ServiceMetadataProvider = SymbolUtils.createValueSymbolBuilder(
+                                    "RegisterServiceMetadata", AwsGoDependency.AWS_MIDDLEWARE).build();
+
+                            Optional<ServiceTrait> serviceTrait = service.getTrait(ServiceTrait.class);
+                            Optional<SigV4Trait> sigV4Trait = service.getTrait(SigV4Trait.class);
+
+                            StringBuilder builder = new StringBuilder();
+                            builder.append("$T{\n");
+                            builder.append("Region: options.Region,\n");
+                            if (serviceTrait.isPresent()) {
+                                ServiceTrait trait = serviceTrait.get();
+                                String sdkIdAsSymbol = trait.getSdkId().toLowerCase().replaceAll("\\s+", "");
+                                builder.append(String.format("ServiceName: \"%s\",\n", trait.getSdkId()));
+                                builder.append(String.format("ServiceID: \"%s\",\n", sdkIdAsSymbol));
+                                // TODO: EndpointID can be different but is not modeled in Smithy.
+                                builder.append(String.format("EndpointID: \"%s\",\n", sdkIdAsSymbol));
+                            }
+                            if (sigV4Trait.isPresent()) {
+                                SigV4Trait trait = sigV4Trait.get();
+                                builder.append(String.format("SigningName: \"%s\",\n", trait.getName()));
+                            }
+                            builder.append(String.format("OperationName: \"%s\",\n", operation.getId().getName()));
+                            builder.append("}");
+
+                            writeAddMiddlewareBefore(writer, INITIALIZE_MIDDLEWARE, ServiceMetadataProvider,
+                                    builder.toString(), stackOperand);
```

Comment:
Is the middleware utilities the reason that need to use string format here to build a string instead of using the writer directly with $T, $L etc?

Would it make sense for a version of writeAddMiddlewareBefore to be defined that takes  Consummable<<GoWriter> ?

Reply:
Yes this is due to the nature of writeAddMiddlewareBefore has been defined. Comments regarding this design should probably be addressed by Shantanu's relevant PRs that are refactoring this area,